```
Error running mint automation
+ mkdir -p 10945-65fb1c6/gopath/src/github.com/minio
+ git -C 10945-65fb1c6/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 10945-65fb1c6/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 10945-65fb1c6/gopath/src/github.com/minio/minio fetch --quiet minio pull/10945/head:pr10945
+ git -C 10945-65fb1c6/gopath/src/github.com/minio/minio checkout --quiet pr10945
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/10945-65fb1c6/gopath
+ make -C 10945-65fb1c6/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/minio/cli v1.22.0
go: downloading github.com/minio/sha256-simd v0.1.1
go: downloading github.com/json-iterator/go v1.1.10
go: downloading github.com/colinmarc/hdfs/v2 v2.1.1
go: downloading github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: downloading google.golang.org/api v0.5.0
go: downloading github.com/minio/minio-go/v7 v7.0.6-0.20201118225257-f6869a5e2a6a
go: downloading gopkg.in/jcmturner/gokrb5.v7 v7.3.0
go: downloading go.etcd.io/etcd/v3 v3.3.0-rc.0.0.20200707003333-58bb8ae09f8e
go: downloading github.com/Azure/azure-storage-blob-go v0.10.0
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/prometheus/client_golang v1.0.0
go: downloading github.com/Azure/azure-pipeline-go v0.2.2
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/cespare/xxhash/v2 v2.1.1
go: downloading github.com/shirou/gopsutil v2.20.9+incompatible
go: downloading github.com/fatih/color v1.9.0
go: downloading golang.org/x/crypto v0.0.0-20200820211705-5c72a883971a
go: downloading golang.org/x/sys v0.0.0-20200915084602-288bc346aa39
go: downloading github.com/beorn7/perks v1.0.0
go: downloading github.com/minio/sio v0.2.1
go: downloading github.com/golang/protobuf v1.4.0
go: downloading github.com/gorilla/handlers v1.4.2
go: downloading github.com/minio/selfupdate v0.3.1
go: downloading github.com/tinylib/msgp v1.1.3
go: downloading github.com/mattn/go-runewidth v0.0.4
go: downloading github.com/klauspost/readahead v1.3.1
go: downloading google.golang.org/protobuf v1.22.0
go: downloading github.com/mattn/go-isatty v0.0.11
go: downloading github.com/klauspost/compress v1.10.3
go: downloading github.com/klauspost/reedsolomon v1.9.9
go: downloading github.com/gorilla/rpc v1.2.0
go: downloading github.com/prometheus/client_model v0.0.0-20190812154241-14fe0d1b01d4
go: downloading github.com/miekg/dns v1.1.8
go: downloading cloud.google.com/go v0.39.0
go: downloading github.com/mattn/go-colorable v0.1.4
go: downloading github.com/prometheus/common v0.4.1
go: downloading github.com/elazarl/go-bindata-assetfs v1.0.0
go: downloading github.com/secure-io/sio-go v0.3.0
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/minio/highwayhash v1.0.0
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/klauspost/cpuid v1.3.1
go: downloading github.com/prometheus/procfs v0.0.8
go: downloading github.com/minio/md5-simd v1.1.0
go: downloading golang.org/x/net v0.0.0-20200904194848-62affa334b73
go: downloading github.com/fatih/structs v1.1.0
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/Shopify/sarama v1.24.1
go: downloading github.com/gogo/protobuf v1.3.1
go: downloading github.com/montanaflynn/stats v0.5.0
go: downloading gopkg.in/ldap.v3 v3.0.3
go: downloading gopkg.in/yaml.v2 v2.2.8
go: downloading github.com/rs/cors v1.7.0
go: downloading gopkg.in/olivere/elastic.v5 v5.0.86
go: downloading gopkg.in/ini.v1 v1.57.0
go: downloading github.com/eapache/go-resiliency v1.1.0
go: downloading github.com/nsqio/go-nsq v1.0.7
go: downloading github.com/philhofer/fwd v1.1.0
go: downloading github.com/googleapis/gax-go/v2 v2.0.4
go: downloading github.com/nats-io/stan.go v0.7.0
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/rcrowley/go-metrics v0.0.0-20181016184325-3113b8401b8a
go: downloading github.com/go-sql-driver/mysql v1.5.0
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/gomodule/redigo v2.0.0+incompatible
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading google.golang.org/grpc v1.26.0
go: downloading go.opencensus.io v0.22.3
go: downloading github.com/hashicorp/vault/api v1.0.4
go: downloading github.com/mitchellh/mapstructure v1.1.2
go: downloading github.com/coredns/coredns v1.4.0
go: downloading gopkg.in/asn1-ber.v1 v1.0.0-20181015200546-f715ec2f112d
go: downloading github.com/hashicorp/hcl v1.0.0
go: downloading github.com/dswarbrick/smart v0.0.0-20190505152634-909a45200d6d
go: downloading github.com/pierrec/lz4 v2.4.0+incompatible
go: downloading github.com/jcmturner/gofork v0.0.0-20190328161633-dc7c13fece03
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/streadway/amqp v0.0.0-20190404075320-75d898a42a94
go: downloading github.com/mailru/easyjson v0.7.1
go: downloading google.golang.org/genproto v0.0.0-20190819201941-24fa4b261c55
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/hashicorp/go-retryablehttp v0.5.4
go: downloading github.com/willf/bloom v2.0.3+incompatible
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/hashicorp/go-cleanhttp v0.5.1
go: downloading github.com/hashicorp/vault/sdk v0.1.13
go: downloading github.com/valyala/tcplisten v0.0.0-20161114210144-ceec8f93295a
go: downloading golang.org/x/oauth2 v0.0.0-20190226205417-e64efc72b421
go: downloading github.com/golang/snappy v0.0.1
go: downloading github.com/modern-go/reflect2 v1.0.1
go: downloading github.com/eclipse/paho.mqtt.golang v1.2.0
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/hashicorp/go-uuid v1.0.1
go: downloading golang.org/x/text v0.3.3
go: downloading github.com/hashicorp/go-rootcerts v1.0.1
go: downloading github.com/willf/bitset v1.1.11
go: downloading gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: downloading github.com/spaolacci/murmur3 v1.1.0
go: downloading go.uber.org/zap v1.14.1
go: downloading github.com/lib/pq v1.8.0
go: downloading github.com/google/uuid v1.1.1
go: downloading github.com/hashicorp/go-sockaddr v1.0.2
go: downloading github.com/coreos/go-semver v0.2.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.1
go: downloading github.com/golang/groupcache v0.0.0-20190702054246-869f871628b6
go: downloading gopkg.in/square/go-jose.v2 v2.3.1
go: downloading go.uber.org/atomic v1.6.0
go: downloading github.com/rs/xid v1.2.1
go: downloading github.com/hashicorp/errwrap v1.0.0
go: downloading golang.org/x/time v0.0.0-20190308202827-9d24e82272b4
go: downloading github.com/minio/simdjson-go v0.1.5
go: downloading github.com/hashicorp/go-multierror v1.0.0
go: downloading github.com/coreos/go-systemd/v22 v22.0.0
go: downloading gopkg.in/jcmturner/goidentity.v3 v3.0.0
go: downloading go.uber.org/multierr v1.5.0
go: downloading github.com/nats-io/nats.go v1.10.0
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading git.apache.org/thrift.git v0.13.0
go: downloading gopkg.in/jcmturner/aescts.v1 v1.0.1
go: downloading github.com/nats-io/jwt v0.3.2
go: downloading github.com/ryanuber/go-glob v1.0.0
go: downloading github.com/nats-io/nkeys v0.1.4
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/klauspost/pgzip v1.2.1
go: downloading gopkg.in/jcmturner/rpc.v1 v1.1.0
go: downloading github.com/tidwall/gjson v1.3.5
go: downloading github.com/tidwall/sjson v1.0.4
go: downloading github.com/tidwall/match v1.0.1
go: downloading github.com/tidwall/pretty v1.0.0
# github.com/minio/minio/cmd
cmd/admin-bucket-handlers.go:167:43: target.URL().HostName undefined (type *url.URL has no field or method HostName, but does have Hostname)
make: *** [build] Error 2

```

--- a/cmd/admin-bucket-handlers.go
+++ b/cmd/admin-bucket-handlers.go
@@ -20,7 +20,6 @@ import (
 	"encoding/json"
 	"io"
 	"io/ioutil"
-	"net"
 	"net/http"
 
 	"github.com/gorilla/mux"
@@ -165,12 +164,8 @@ func (a adminAPIHandlers) SetRemoteTargetHandler(w http.ResponseWriter, r *http.
 		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErrWithErr(ErrAdminConfigBadJSON, err), r.URL)
 		return
 	}
-	host, port, err := net.SplitHostPort(target.Endpoint)
-	if err != nil {
-		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErrWithErr(ErrAdminConfigBadJSON, err), r.URL)
-		return
-	}
-	sameTarget, _ := isLocalHost(host, port, globalMinioPort)
+
+	sameTarget, _ := isLocalHost(target.URL().Hostname(), target.URL().Port(), globalMinioPort)
 	if sameTarget && bucket == target.TargetBucket {
 		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrBucketRemoteIdenticalToSource), r.URL)
 		return

--- a/cmd/api-datatypes.go
+++ b/cmd/api-datatypes.go
@@ -35,6 +35,8 @@ type DeletedObject struct {
 	DeleteMarkerMTime time.Time `xml:"DeleteMarkerMTime,omitempty"`
 	// Status of versioned delete (of object or DeleteMarker)
 	VersionPurgeStatus VersionPurgeStatusType `xml:"VersionPurgeStatus,omitempty"`
+	// PurgeTransitioned is nonempty if object is in transition tier
+	PurgeTransitioned string `xml:"PurgeTransitioned,omitempty"`
 }
 
 // ObjectToDelete carries key name for the object to delete.
@@ -47,6 +49,8 @@ type ObjectToDelete struct {
 	VersionPurgeStatus VersionPurgeStatusType `xml:"VersionPurgeStatus"`
 	// Version ID of delete marker
 	DeleteMarkerVersionID string `xml:"DeleteMarkerVersionId"`
+	// PurgeTransitioned is nonempty if object is in transition tier
+	PurgeTransitioned string `xml:"PurgeTransitioned"`
 }
 
 // createBucketConfiguration container for bucket configuration request from client.

--- a/cmd/bucket-handlers.go
+++ b/cmd/bucket-handlers.go
@@ -37,6 +37,7 @@ import (
 	"github.com/minio/minio/cmd/crypto"
 	xhttp "github.com/minio/minio/cmd/http"
 	"github.com/minio/minio/cmd/logger"
+	"github.com/minio/minio/pkg/bucket/lifecycle"
 	objectlock "github.com/minio/minio/pkg/bucket/object/lock"
 	"github.com/minio/minio/pkg/bucket/policy"
 	"github.com/minio/minio/pkg/bucket/replication"
@@ -404,8 +405,18 @@ func (api objectAPIHandlers) DeleteMultipleObjectsHandler(w http.ResponseWriter,
 	if api.CacheAPI() != nil {
 		getObjectInfoFn = api.CacheAPI().GetObjectInfo
 	}
+	var (
+		hasLockEnabled, hasLifecycleConfig bool
+		goi                                ObjectInfo
+		gerr                               error
+	)
 	replicateDeletes := hasReplicationRules(ctx, bucket, deleteObjects.Objects)
-
+	if rcfg, _ := globalBucketObjectLockSys.Get(bucket); rcfg.LockEnabled {
+		hasLockEnabled = true
+	}
+	if _, err := globalBucketMetadataSys.GetLifecycleConfig(bucket); err == nil {
+		hasLifecycleConfig = true
+	}
 	dErrs := make([]DeleteError, len(deleteObjects.Objects))
 	for index, object := range deleteObjects.Objects {
 		if apiErrCode := checkRequestAuthType(ctx, r, policy.DeleteObjectAction, bucket, object.ObjectName); apiErrCode != ErrNone {
@@ -422,10 +433,33 @@ func (api objectAPIHandlers) DeleteMultipleObjectsHandler(w http.ResponseWriter,
 			}
 			continue
 		}
-
+		if replicateDeletes || hasLockEnabled || hasLifecycleConfig {
+			goi, gerr = getObjectInfoFn(ctx, bucket, object.ObjectName, ObjectOptions{
+				VersionID: object.VersionID,
+			})
+		}
+		if hasLifecycleConfig && gerr == nil {
+			object.PurgeTransitioned = goi.TransitionStatus
+		}
+		if replicateDeletes {
+			delMarker, replicate := checkReplicateDelete(ctx, bucket, ObjectToDelete{
+				ObjectName: object.ObjectName,
+				VersionID:  object.VersionID,
+			}, goi, gerr)
+			if replicate {
+				if object.VersionID != "" {
+					object.VersionPurgeStatus = Pending
+					if delMarker {
+						object.DeleteMarkerVersionID = object.VersionID
+					}
+				} else {
+					object.DeleteMarkerReplicationStatus = string(replication.Pending)
+				}
+			}
+		}
 		if object.VersionID != "" {
-			if rcfg, _ := globalBucketObjectLockSys.Get(bucket); rcfg.LockEnabled {
-				if apiErrCode := enforceRetentionBypassForDelete(ctx, r, bucket, object, getObjectInfoFn); apiErrCode != ErrNone {
+			if hasLockEnabled {
+				if apiErrCode := enforceRetentionBypassForDelete(ctx, r, bucket, object, goi, gerr); apiErrCode != ErrNone {
 					apiErr := errorCodes.ToAPIErr(apiErrCode)
 					dErrs[index] = DeleteError{
 						Code:      apiErr.Code,
@@ -440,22 +474,6 @@ func (api objectAPIHandlers) DeleteMultipleObjectsHandler(w http.ResponseWriter,
 
 		// Avoid duplicate objects, we use map to filter them out.
 		if _, ok := objectsToDelete[object]; !ok {
-			if replicateDeletes {
-				delMarker, replicate := checkReplicateDelete(ctx, getObjectInfoFn, bucket, ObjectToDelete{
-					ObjectName: object.ObjectName,
-					VersionID:  object.VersionID,
-				})
-				if replicate {
-					if object.VersionID != "" {
-						object.VersionPurgeStatus = Pending
-						if delMarker {
-							object.DeleteMarkerVersionID = object.VersionID
-						}
-					} else {
-						object.DeleteMarkerReplicationStatus = string(replication.Pending)
-					}
-				}
-			}
 			objectsToDelete[object] = index
 		}
 	}
@@ -475,24 +493,17 @@ func (api objectAPIHandlers) DeleteMultipleObjectsHandler(w http.ResponseWriter,
 		Versioned:        globalBucketVersioningSys.Enabled(bucket),
 		VersionSuspended: globalBucketVersioningSys.Suspended(bucket),
 	})
-
 	deletedObjects := make([]DeletedObject, len(deleteObjects.Objects))
 	for i := range errs {
-		var dindex int
-		if replicateDeletes {
-			dindex = objectsToDelete[ObjectToDelete{
-				ObjectName:                    dObjects[i].ObjectName,
-				VersionID:                     dObjects[i].VersionID,
-				DeleteMarkerVersionID:         dObjects[i].DeleteMarkerVersionID,
-				VersionPurgeStatus:            dObjects[i].VersionPurgeStatus,
-				DeleteMarkerReplicationStatus: dObjects[i].DeleteMarkerReplicationStatus,
-			}]
-		} else {
-			dindex = objectsToDelete[ObjectToDelete{
-				ObjectName: dObjects[i].ObjectName,
-				VersionID:  dObjects[i].VersionID,
-			}]
-		}
+
+		dindex := objectsToDelete[ObjectToDelete{
+			ObjectName:                    dObjects[i].ObjectName,
+			VersionID:                     dObjects[i].VersionID,
+			DeleteMarkerVersionID:         dObjects[i].DeleteMarkerVersionID,
+			VersionPurgeStatus:            dObjects[i].VersionPurgeStatus,
+			DeleteMarkerReplicationStatus: dObjects[i].DeleteMarkerReplicationStatus,
+			PurgeTransitioned:             dObjects[i].PurgeTransitioned,
+		}]
 		if errs[i] == nil || isErrObjectNotFound(errs[i]) || isErrVersionNotFound(errs[i]) {
 			if replicateDeletes {
 				dObjects[i].DeleteMarkerReplicationStatus = deleteList[i].DeleteMarkerReplicationStatus
@@ -532,6 +543,19 @@ func (api objectAPIHandlers) DeleteMultipleObjectsHandler(w http.ResponseWriter,
 				})
 			}
 		}
+
+		if hasLifecycleConfig && dobj.PurgeTransitioned == lifecycle.TransitionComplete { // clean up transitioned tier
+			action := lifecycle.DeleteAction
+			if dobj.VersionID != "" {
+				action = lifecycle.DeleteVersionAction
+			}
+			deleteTransitionedObject(ctx, newObjectLayerFn(), bucket, dobj.ObjectName, lifecycle.ObjectOpts{
+				Name:         dobj.ObjectName,
+				VersionID:    dobj.VersionID,
+				DeleteMarker: dobj.DeleteMarker,
+			}, action, true)
+		}
+
 	}
 	// Notify deleted event for objects.
 	for _, dobj := range deletedObjects {

--- a/cmd/bucket-lifecycle.go
+++ b/cmd/bucket-lifecycle.go
@@ -111,7 +111,9 @@ func (t *transitionState) addWorker(ctx context.Context, objectAPI ObjectLayer) 
 				if !ok {
 					return
 				}
-				transitionObject(ctx, objectAPI, oi)
+				if err := transitionObject(ctx, objectAPI, oi); err != nil {
+					logger.LogIf(ctx, err)
+				}
 			}
 		}
 	}()
@@ -244,8 +246,9 @@ func putTransitionOpts(objInfo ObjectInfo) (putOpts miniogo.PutObjectOptions) {
 // handle deletes of transitioned objects or object versions when one of the following is true:
 // 1. temporarily restored copies of objects (restored with the PostRestoreObject API) expired.
 // 2. life cycle expiry date is met on the object.
-func deleteTransitionedObject(ctx context.Context, objectAPI ObjectLayer, bucket, object string, objInfo ObjectInfo, lcOpts lifecycle.ObjectOpts, action lifecycle.Action) error {
-	if objInfo.TransitionStatus == "" {
+// 3. Object is removed through DELETE api call
+func deleteTransitionedObject(ctx context.Context, objectAPI ObjectLayer, bucket, object string, lcOpts lifecycle.ObjectOpts, action lifecycle.Action, isDeleteTierOnly bool) error {
+	if lcOpts.TransitionStatus == "" && !isDeleteTierOnly {
 		return nil
 	}
 	lc, err := globalLifecycleSys.Get(bucket)
@@ -263,28 +266,37 @@ func deleteTransitionedObject(ctx context.Context, objectAPI ObjectLayer, bucket
 
 	var opts ObjectOptions
 	opts.Versioned = globalBucketVersioningSys.Enabled(bucket)
-	opts.VersionID = objInfo.VersionID
+	opts.VersionID = lcOpts.VersionID
 	switch action {
 	case lifecycle.DeleteRestoredAction, lifecycle.DeleteRestoredVersionAction:
 		// delete locally restored copy of object or object version
 		// from the source, while leaving metadata behind. The data on
 		// transitioned tier lies untouched and still accessible
-		opts.TransitionStatus = objInfo.TransitionStatus
+		opts.TransitionStatus = lcOpts.TransitionStatus
 		_, err = objectAPI.DeleteObject(ctx, bucket, object, opts)
 		return err
 	case lifecycle.DeleteAction, lifecycle.DeleteVersionAction:
 		// When an object is past expiry, delete the data from transitioned tier and
 		// metadata from source
-		if err := tgt.RemoveObject(ctx, arn.Bucket, object, miniogo.RemoveObjectOptions{VersionID: objInfo.VersionID}); err != nil {
+		if err := tgt.RemoveObject(context.Background(), arn.Bucket, object, miniogo.RemoveObjectOptions{VersionID: lcOpts.VersionID}); err != nil {
 			logger.LogIf(ctx, err)
+		}
+
+		if isDeleteTierOnly {
+			return nil
 		}
 		_, err = objectAPI.DeleteObject(ctx, bucket, object, opts)
 		if err != nil {
 			return err
 		}
 		eventName := event.ObjectRemovedDelete
-		if objInfo.DeleteMarker {
+		if lcOpts.DeleteMarker {
 			eventName = event.ObjectRemovedDeleteMarkerCreated
+		}
+		objInfo := ObjectInfo{
+			Name:         object,
+			VersionID:    lcOpts.VersionID,
+			DeleteMarker: lcOpts.DeleteMarker,
 		}
 		// Notify object deleted event.
 		sendEvent(eventArgs{

--- a/cmd/bucket-object-lock.go
+++ b/cmd/bucket-object-lock.go
@@ -82,7 +82,7 @@ func enforceRetentionForDeletion(ctx context.Context, objInfo ObjectInfo) (locke
 // For objects in "Governance" mode, overwrite is allowed if a) object retention date is past OR
 // governance bypass headers are set and user has governance bypass permissions.
 // Objects in "Compliance" mode can be overwritten only if retention date is past.
-func enforceRetentionBypassForDelete(ctx context.Context, r *http.Request, bucket string, object ObjectToDelete, getObjectInfoFn GetObjectInfoFn) APIErrorCode {
+func enforceRetentionBypassForDelete(ctx context.Context, r *http.Request, bucket string, object ObjectToDelete, oi ObjectInfo, gerr error) APIErrorCode {
 	opts, err := getOpts(ctx, r, bucket, object.ObjectName)
 	if err != nil {
 		return toAPIErrorCode(ctx, err)
@@ -90,8 +90,7 @@ func enforceRetentionBypassForDelete(ctx context.Context, r *http.Request, bucke
 
 	opts.VersionID = object.VersionID
 
-	oi, err := getObjectInfoFn(ctx, bucket, object.ObjectName, opts)
-	if err != nil {
+	if gerr != nil { // error from GetObjectInfo
 		switch err.(type) {
 		case MethodNotAllowed: // This happens usually for a delete marker
 			if oi.DeleteMarker {

--- a/cmd/bucket-replication.go
+++ b/cmd/bucket-replication.go
@@ -144,15 +144,14 @@ func hasReplicationRules(ctx context.Context, bucket string, objects []ObjectToD
 }
 
 // returns whether object version is a deletemarker and if object qualifies for replication
-func checkReplicateDelete(ctx context.Context, getObjectInfoFn GetObjectInfoFn, bucket string, dobj ObjectToDelete) (dm, replicate bool) {
+func checkReplicateDelete(ctx context.Context, bucket string, dobj ObjectToDelete, oi ObjectInfo, gerr error) (dm, replicate bool) {
 	rcfg, err := getReplicationConfig(ctx, bucket)
 	if err != nil || rcfg == nil {
 		return false, false
 	}
-	oi, err := getObjectInfoFn(ctx, bucket, dobj.ObjectName, ObjectOptions{VersionID: dobj.VersionID})
 	// when incoming delete is removal of a delete marker( a.k.a versioned delete),
 	// GetObjectInfo returns extra information even though it returns errFileNotFound
-	if err != nil {
+	if gerr != nil {
 		validReplStatus := false
 		switch oi.ReplicationStatus {
 		case replication.Pending, replication.Complete, replication.Failed:
@@ -469,6 +468,7 @@ func replicateObject(ctx context.Context, objInfo ObjectInfo, objectAPI ObjectLa
 		dstOpts := miniogo.PutObjectOptions{Internal: miniogo.AdvancedPutOptions{SourceVersionID: objInfo.VersionID}}
 		_, err = tgt.CopyObject(ctx, dest.Bucket, object, dest.Bucket, object, getCopyObjMetadata(objInfo, dest), dstOpts)
 	}
+
 	r.Close()
 	if err != nil {
 		replicationStatus = replication.Failed

--- a/cmd/bucket-targets.go
+++ b/cmd/bucket-targets.go
@@ -18,6 +18,7 @@ package cmd
 
 import (
 	"context"
+	"encoding/hex"
 	"net/http"
 	"strings"
 	"sync"
@@ -28,6 +29,7 @@ import (
 	"github.com/minio/minio-go/v7/pkg/credentials"
 	"github.com/minio/minio/pkg/bucket/versioning"
 	"github.com/minio/minio/pkg/madmin"
+	sha256 "github.com/minio/sha256-simd"
 )
 
 // BucketTargetSys represents bucket targets subsystem
@@ -371,11 +373,21 @@ func (sys *BucketTargetSys) getRemoteARN(bucket string, target *madmin.BucketTar
 	if !madmin.ServiceType(target.Type).IsValid() {
 		return ""
 	}
+	return generateARN(target)
+}
+
+// generate ARN that is unique to this target type
+func generateARN(t *madmin.BucketTarget) string {
+	hash := sha256.New()
+	hash.Write([]byte(t.Type))
+	hash.Write([]byte(t.Region))
+	hash.Write([]byte(t.TargetBucket))
+	hashSum := hex.EncodeToString(hash.Sum(nil))
 	arn := madmin.ARN{
-		Type:   target.Type,
-		ID:     mustGetUUID(),
-		Region: target.Region,
-		Bucket: target.TargetBucket,
+		Type:   t.Type,
+		ID:     hashSum,
+		Region: t.Region,
+		Bucket: t.TargetBucket,
 	}
 	return arn.String()
 }

--- a/cmd/bucket-targets.go
+++ b/cmd/bucket-targets.go
@@ -351,7 +351,7 @@ func (sys *BucketTargetSys) getRemoteTargetClient(tcfg *madmin.BucketTarget) (*m
 		getRemoteTargetInstanceTransport = newGatewayHTTPTransport(1 * time.Hour)
 	})
 
-	core, err := miniogo.NewCore(tcfg.Endpoint, &miniogo.Options{
+	core, err := miniogo.NewCore(tcfg.URL().Host, &miniogo.Options{
 		Creds:     creds,
 		Secure:    tcfg.Secure,
 		Transport: getRemoteTargetInstanceTransport,
@@ -366,7 +366,7 @@ func (sys *BucketTargetSys) getRemoteARN(bucket string, target *madmin.BucketTar
 	}
 	tgts := sys.targetsMap[bucket]
 	for _, tgt := range tgts {
-		if tgt.Type == target.Type && tgt.TargetBucket == target.TargetBucket && target.URL() == tgt.URL() {
+		if tgt.Type == target.Type && tgt.TargetBucket == target.TargetBucket && target.URL().String() == tgt.URL().String() {
 			return tgt.Arn
 		}
 	}

--- a/cmd/data-crawler.go
+++ b/cmd/data-crawler.go
@@ -772,7 +772,7 @@ func (i *crawlItem) applyActions(ctx context.Context, o ObjectLayer, meta action
 		return 0
 	}
 	if obj.TransitionStatus != "" {
-		if err := deleteTransitionedObject(ctx, o, i.bucket, i.objectPath(), obj, lcOpts, action); err != nil {
+		if err := deleteTransitionedObject(ctx, o, i.bucket, i.objectPath(), lcOpts, action, false); err != nil {
 			logger.LogIf(ctx, err)
 			return size
 		}

--- a/cmd/erasure-object.go
+++ b/cmd/erasure-object.go
@@ -905,6 +905,7 @@ func (er erasureObjects) DeleteObjects(ctx context.Context, bucket string, objec
 					DeleteMarkerReplicationStatus: versions[objIndex].DeleteMarkerReplicationStatus,
 					ObjectName:                    versions[objIndex].Name,
 					VersionPurgeStatus:            versions[objIndex].VersionPurgeStatus,
+					PurgeTransitioned:             objects[objIndex].PurgeTransitioned,
 				}
 			} else {
 				dobjects[objIndex] = DeletedObject{
@@ -912,6 +913,7 @@ func (er erasureObjects) DeleteObjects(ctx context.Context, bucket string, objec
 					VersionID:                     versions[objIndex].VersionID,
 					VersionPurgeStatus:            versions[objIndex].VersionPurgeStatus,
 					DeleteMarkerReplicationStatus: versions[objIndex].DeleteMarkerReplicationStatus,
+					PurgeTransitioned:             objects[objIndex].PurgeTransitioned,
 				}
 			}
 		}

--- a/cmd/object-handlers.go
+++ b/cmd/object-handlers.go
@@ -2714,7 +2714,24 @@ func (api objectAPIHandlers) DeleteObjectHandler(w http.ResponseWriter, r *http.
 		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
 		return
 	}
-	_, replicateDel := checkReplicateDelete(ctx, getObjectInfo, bucket, ObjectToDelete{ObjectName: object, VersionID: opts.VersionID})
+	var (
+		hasLockEnabled, hasLifecycleConfig bool
+		goi                                ObjectInfo
+		gerr                               error
+	)
+	replicateDeletes := hasReplicationRules(ctx, bucket, []ObjectToDelete{{ObjectName: object, VersionID: opts.VersionID}})
+	if rcfg, _ := globalBucketObjectLockSys.Get(bucket); rcfg.LockEnabled {
+		hasLockEnabled = true
+	}
+	if _, err := globalBucketMetadataSys.GetLifecycleConfig(bucket); err == nil {
+		hasLifecycleConfig = true
+	}
+	if replicateDeletes || hasLockEnabled || hasLifecycleConfig {
+		goi, gerr = getObjectInfo(ctx, bucket, object, ObjectOptions{
+			VersionID: opts.VersionID,
+		})
+	}
+	_, replicateDel := checkReplicateDelete(ctx, bucket, ObjectToDelete{ObjectName: object, VersionID: opts.VersionID}, goi, gerr)
 	if replicateDel {
 		if opts.VersionID != "" {
 			opts.VersionPurgeStatus = Pending
@@ -2722,7 +2739,7 @@ func (api objectAPIHandlers) DeleteObjectHandler(w http.ResponseWriter, r *http.
 			opts.DeleteMarkerReplicationStatus = string(replication.Pending)
 		}
 	}
-	replicaDel := false
+	vID := opts.VersionID
 	if r.Header.Get(xhttp.AmzBucketReplicationStatus) == replication.Replica.String() {
 		// check if replica has permission to be deleted.
 		if apiErrCode := checkRequestAuthType(ctx, r, policy.ReplicateDeleteAction, bucket, object); apiErrCode != ErrNone {
@@ -2730,20 +2747,19 @@ func (api objectAPIHandlers) DeleteObjectHandler(w http.ResponseWriter, r *http.
 			return
 		}
 		opts.DeleteMarkerReplicationStatus = replication.Replica.String()
-		replicaDel = true
+		if opts.VersionPurgeStatus.Empty() {
+			// opts.VersionID holds delete marker version ID to replicate and not yet present on disk
+			vID = ""
+		}
 	}
-	vID := opts.VersionID
-	if replicaDel && opts.VersionPurgeStatus.Empty() {
-		// opts.VersionID holds delete marker version ID to replicate and not yet present on disk
-		vID = ""
-	}
+
 	apiErr := ErrNone
 	if rcfg, _ := globalBucketObjectLockSys.Get(bucket); rcfg.LockEnabled {
 		if vID != "" {
 			apiErr = enforceRetentionBypassForDelete(ctx, r, bucket, ObjectToDelete{
 				ObjectName: object,
 				VersionID:  vID,
-			}, getObjectInfo)
+			}, goi, gerr)
 			if apiErr != ErrNone && apiErr != ErrNoSuchKey {
 				writeErrorResponse(ctx, w, errorCodes.ToAPIErr(apiErr), r.URL, guessIsBrowserReq(r))
 				return
@@ -2788,6 +2804,20 @@ func (api objectAPIHandlers) DeleteObjectHandler(w http.ResponseWriter, r *http.
 			},
 			Bucket: bucket,
 		})
+	}
+	if goi.TransitionStatus == lifecycle.TransitionComplete { // clean up transitioned tier
+		action := lifecycle.DeleteAction
+		if goi.VersionID != "" {
+			action = lifecycle.DeleteVersionAction
+		}
+		deleteTransitionedObject(ctx, newObjectLayerFn(), bucket, object, lifecycle.ObjectOpts{
+			Name:             object,
+			UserTags:         goi.UserTags,
+			VersionID:        goi.VersionID,
+			DeleteMarker:     goi.DeleteMarker,
+			TransitionStatus: goi.TransitionStatus,
+			IsLatest:         goi.IsLatest,
+		}, action, true)
 	}
 	setPutObjHeaders(w, objInfo, true)
 	writeSuccessNoContent(w)

--- a/cmd/web-handlers.go
+++ b/cmd/web-handlers.go
@@ -47,6 +47,7 @@ import (
 	xhttp "github.com/minio/minio/cmd/http"
 	"github.com/minio/minio/cmd/logger"
 	"github.com/minio/minio/pkg/auth"
+	"github.com/minio/minio/pkg/bucket/lifecycle"
 	objectlock "github.com/minio/minio/pkg/bucket/object/lock"
 	"github.com/minio/minio/pkg/bucket/policy"
 	"github.com/minio/minio/pkg/bucket/replication"
@@ -737,14 +738,24 @@ next:
 					return toJSONError(ctx, errAccessDenied)
 				}
 			}
-			_, replicateDel := checkReplicateDelete(ctx, getObjectInfoFn, args.BucketName, ObjectToDelete{ObjectName: objectName})
-			if replicateDel {
-				opts.DeleteMarkerReplicationStatus = string(replication.Pending)
-				opts.DeleteMarker = true
+			var (
+				replicateDel, hasLifecycleConfig bool
+				goi                              ObjectInfo
+				gerr                             error
+			)
+			if _, err := globalBucketMetadataSys.GetLifecycleConfig(args.BucketName); err == nil {
+				hasLifecycleConfig = true
+			}
+			if hasReplicationRules(ctx, args.BucketName, []ObjectToDelete{{ObjectName: objectName}}) || hasLifecycleConfig {
+				goi, gerr = getObjectInfoFn(ctx, args.BucketName, objectName, opts)
+				if _, replicateDel = checkReplicateDelete(ctx, args.BucketName, ObjectToDelete{ObjectName: objectName}, goi, gerr); replicateDel {
+					opts.DeleteMarkerReplicationStatus = string(replication.Pending)
+					opts.DeleteMarker = true
+				}
 			}
 
 			oi, err := deleteObject(ctx, objectAPI, web.CacheAPI(), args.BucketName, objectName, nil, r, opts)
-			if replicateDel {
+			if replicateDel && err == nil {
 				globalReplicationState.queueReplicaDeleteTask(DeletedObjectVersionInfo{
 					DeletedObject: DeletedObject{
 						ObjectName:                    objectName,
@@ -756,6 +767,19 @@ next:
 					},
 					Bucket: args.BucketName,
 				})
+			}
+			if goi.TransitionStatus == lifecycle.TransitionComplete && err == nil && goi.VersionID == "" {
+				action := lifecycle.DeleteAction
+				if goi.VersionID != "" {
+					action = lifecycle.DeleteVersionAction
+				}
+				deleteTransitionedObject(ctx, newObjectLayerFn(), args.BucketName, objectName, lifecycle.ObjectOpts{
+					Name:         objectName,
+					UserTags:     goi.UserTags,
+					VersionID:    goi.VersionID,
+					DeleteMarker: goi.DeleteMarker,
+					IsLatest:     goi.IsLatest,
+				}, action, true)
 			}
 
 			logger.LogIf(ctx, err)
@@ -829,7 +853,7 @@ next:
 					}
 				}
 				// since versioned delete is not available on web browser, yet - this is a simple DeleteMarker replication
-				_, replicateDel := checkReplicateDelete(ctx, getObjectInfoFn, args.BucketName, ObjectToDelete{ObjectName: obj.Name})
+				_, replicateDel := checkReplicateDelete(ctx, args.BucketName, ObjectToDelete{ObjectName: obj.Name}, obj, nil)
 				objToDel := ObjectToDelete{ObjectName: obj.Name}
 				if replicateDel {
 					objToDel.DeleteMarkerReplicationStatus = string(replication.Pending)

--- a/cmd/xl-storage-format-v2.go
+++ b/cmd/xl-storage-format-v2.go
@@ -510,6 +510,9 @@ func (z *xlMetaV2) DeleteVersion(fi FileInfo) (string, bool, error) {
 				if updateVersion {
 					delete(z.Versions[i].DeleteMarker.MetaSys, xhttp.AmzBucketReplicationStatus)
 					delete(z.Versions[i].DeleteMarker.MetaSys, VersionPurgeStatusKey)
+					if z.Versions[i].DeleteMarker.MetaSys == nil {
+						z.Versions[i].DeleteMarker.MetaSys = make(map[string][]byte)
+					}
 					if fi.DeleteMarkerReplicationStatus != "" {
 						z.Versions[i].DeleteMarker.MetaSys[xhttp.AmzBucketReplicationStatus] = []byte(fi.DeleteMarkerReplicationStatus)
 					}

--- a/cmd/xl-storage-format-v2.go
+++ b/cmd/xl-storage-format-v2.go
@@ -508,11 +508,11 @@ func (z *xlMetaV2) DeleteVersion(fi FileInfo) (string, bool, error) {
 		case DeleteType:
 			if bytes.Equal(version.DeleteMarker.VersionID[:], uv[:]) {
 				if updateVersion {
-					delete(z.Versions[i].DeleteMarker.MetaSys, xhttp.AmzBucketReplicationStatus)
-					delete(z.Versions[i].DeleteMarker.MetaSys, VersionPurgeStatusKey)
-					if z.Versions[i].DeleteMarker.MetaSys == nil {
+					if len(z.Versions[i].DeleteMarker.MetaSys) == 0 {
 						z.Versions[i].DeleteMarker.MetaSys = make(map[string][]byte)
 					}
+					delete(z.Versions[i].DeleteMarker.MetaSys, xhttp.AmzBucketReplicationStatus)
+					delete(z.Versions[i].DeleteMarker.MetaSys, VersionPurgeStatusKey)
 					if fi.DeleteMarkerReplicationStatus != "" {
 						z.Versions[i].DeleteMarker.MetaSys[xhttp.AmzBucketReplicationStatus] = []byte(fi.DeleteMarkerReplicationStatus)
 					}

--- a/pkg/madmin/remote-target-commands.go
+++ b/pkg/madmin/remote-target-commands.go
@@ -118,20 +118,15 @@ func (t *BucketTarget) Clone() BucketTarget {
 }
 
 // URL returns target url
-func (t BucketTarget) URL() string {
+func (t BucketTarget) URL() *url.URL {
 	scheme := "http"
 	if t.Secure {
 		scheme = "https"
 	}
-	urlStr := fmt.Sprintf("%s://%s", scheme, t.Endpoint)
-	u, err := url.Parse(urlStr)
-	if err != nil {
-		return urlStr
+	return &url.URL{
+		Scheme: scheme,
+		Host:   t.Endpoint,
 	}
-	if u.Port() == "80" || u.Port() == "443" {
-		u.Host = u.Hostname()
-	}
-	return u.String()
 }
 
 // Empty returns true if struct is empty.

